File: frontend/src/api/results.ts
import apiClient from './client'

export interface TimeSeriesDatapoint {
  date: string
  value: number | null
  ref_low: number | null
  ref_high: number | null
}

export interface TimeSeriesSeries {
  analyte_id: string
  analyte_name: string
  unit: string
  datapoints: TimeSeriesDatapoint[]
}

export interface TimeSeriesResponse {
  series: TimeSeriesSeries[]
}

export const resultsAPI = {
  getTimeSeries: async (
    familyMemberId: string,
    analyteIds?: string[],
    dateFrom?: string,
    dateTo?: string
  ): Promise<TimeSeriesResponse> => {
    const params = new URLSearchParams()
    params.append('family_member_id', familyMemberId)
    if (analyteIds?.length) {
      analyteIds.forEach((id) => params.append('analyte_id', id))
    }
    if (dateFrom) params.append('date_from', dateFrom)
    if (dateTo) params.append('date_to', dateTo)

    const response = await apiClient.get(`/results/timeseries?${params.toString()}`)
    return response.data
  },
}

File: backend/app/schemas/results.py
"""Result schemas for API responses."""

from datetime import datetime
from pydantic import BaseModel


class TestResultSchema(BaseModel):
    """Single test result."""

    id: str
    analyte_id: str
    raw_name: str
    canonical_value: float | None
    canonical_unit: str
    report_date: datetime | None
    is_out_of_range: bool

    class Config:
        from_attributes = True


class TimeSeriesDatapoint(BaseModel):
    """Single datapoint in a time-series."""

    date: str
    value: float | None
    ref_low: float | None
    ref_high: float | None
    lab_name: str | None


class TimeSeriesSeries(BaseModel):
    """Time-series data for a single analyte."""

    analyte_id: str
    analyte_name: str
    unit: str
    category: str | None
    ref_low: float | None
    ref_high: float | None
    datapoints: list[TimeSeriesDatapoint]


class TimeSeriesResponseSchema(BaseModel):
    """Response for time-series endpoint."""

    series: list[TimeSeriesSeries]
